vault/external_key_registry.go

```diff
@@ -0,0 +1,628 @@
+	externalKeyRegistryKeyPrefix = "keys/"
+
+	// externalKeyRegistryConfigPrefix is the storage prefix for config entries.
+	externalKeyRegistryConfigPrefix = "configs/"
+)
+
+// ExternalKeyRegistry provides durable storage of config and key mappings and
+// uses these to expose KMS operations to mounts via [logical.ExternalKey].
+type ExternalKeyRegistry struct {
+	core   *Core
+	logger hclog.Logger
+
+	// storageLock ensures consistency both within and between configs and keys.
+	// In the unlikely event this ever becomes a bottleneck, we can shard this
+	// lock per config via locksutil.
+	storageLock sync.RWMutex
+}
+
+// ExternalKey is the storage representation of a key.
+type ExternalKey struct {
+	// KMS-specific key parameters.
+	Values map[string]string `json:"values"`
+
+	// List of mount paths of plugins that may use this key.
+	// Paths are relative to the key's namespace.
+	Grants []string `json:"grants"`
+}
+
+// ExternalKeyConfig is the storage representation a config.
+type ExternalKeyConfig struct {
+	// Type of KMS. May be user-defined via the server config.
+	Type string `json:"type"`
+
+	// Name of a config in the parent namespace to inherit into this config's
+	// namespace. Setting this field requires that Type and Values are empty.
+	Inherits string `json:"inherits"`
+
+	// KMS-specific client parameters.
+	Values map[string]string `json:"values"`
+}
+
+// NewExternalKeyRegistry creates a new [ExternalKeyRegistry].
+func NewExternalKeyRegistry(core *Core, logger hclog.Logger) *ExternalKeyRegistry {
+	return &ExternalKeyRegistry{
+		core:   core,
+		logger: logger,
+	}
+}
+
+// setupExternalKeys initializes [Core.externalKeys].
+func (c *Core) setupExternalKeys() error {
+	logger := c.baseLogger.Named("external-keys")
+	c.AddLogger(logger)
+	c.externalKeys = NewExternalKeyRegistry(c, logger)
+	return nil
+}
+
+// teardownExternalKeys finalizes and zeroes [Core.externalKeys].
+func (c *Core) teardownExternalKeys() error {
+	if c.externalKeys == nil {
+		return nil
+	}
+	c.externalKeys.Finalize()
+	c.externalKeys = nil
+	return nil
+}
+
+// storageViewForRequest returns the registry's root storage view based on the
+// storage field of a [logical.Request], which should automatically be routed to
+// the system barrier by the system backend.
+func (r *ExternalKeyRegistry) storageViewForRequest(req *logical.Request) logical.Storage {
+	return logical.NewStorageView(req.Storage, externalKeyRegistrySubPath)
+}
+
+// configPath returns the storage path for a config.
+func (r *ExternalKeyRegistry) configPath(name string) string {
+	return path.Join(externalKeyRegistryConfigPrefix, name)
+}
+
+// configListPath returns the storage path to list configs at.
+func (r *ExternalKeyRegistry) configListPath() string {
+	return externalKeyRegistryConfigPrefix
+}
+
+// keyPath returns the storage path for a key.
+func (r *ExternalKeyRegistry) keyPath(name string) string {
```

Comment:
Given keys are under configs in the API definition (I thought?), I'm surprised `config` isn't a parameter here?

Reply:
Ah I see. That's fine I suppose.